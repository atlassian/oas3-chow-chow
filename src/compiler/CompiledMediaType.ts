import { MediaTypeObject } from 'openapi3-ts';
import CompiledSchema from './CompiledSchema';
import ChowError from '../error';

export default class CompiledMediaType {
  private name: string;
  private compiledSchema: CompiledSchema;

  constructor(name: string, mediaType: MediaTypeObject) {
    this.name = name;
    this.compiledSchema = new CompiledSchema(mediaType.schema || {},
      {},
      { schemaContext: 'response' });
  }

  public validate(value: any) {
    try {
      this.compiledSchema.validate(value);
      return value;
    } catch(e) {
      throw new ChowError('Schema validation error', { in: `media-type:${this.name}`, rawErrors: e });
    }
  }

  static extractMediaType(contentType: string | undefined): string | undefined {
    if (!contentType) {
      return;
    }

    return contentType.split(';')[0];
  }
}
