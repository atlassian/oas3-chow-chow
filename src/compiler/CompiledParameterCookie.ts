import { ParameterObject, SchemaObject } from 'openapi3-ts';
import CompiledSchema from './CompiledSchema';
import ChowError from '../error';
import { ChowOptions } from '..';

export default class CompiledParameterCookie {
  private compiledSchema: CompiledSchema;
  private cookieSchema: SchemaObject = {
    type: 'object',
    properties: {},
    required: []
  };

  constructor(parameters: ParameterObject[], options: Partial<ChowOptions>) {
    for (const parameter of parameters) {
      this.cookieSchema.properties![parameter.name] = parameter.schema || {};
      if (parameter.required) {
        this.cookieSchema.required!.push(parameter.name);
      }
    }

    this.compiledSchema = new CompiledSchema(this.cookieSchema, options.cookieAjvOptions);
  }

  /**
   * If there is no query passed in, we make it an empty object
   */
  public validate(value: any = {}) {
    try {
      this.compiledSchema.validate(value);
      return value;
    } catch(e) {
      throw new ChowError('Schema validation error', { in: 'cookie', rawErrors: e });
    }
  }
}
