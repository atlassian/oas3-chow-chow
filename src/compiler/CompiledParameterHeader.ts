import { ParameterObject, SchemaObject } from 'openapi3-ts';
import CompiledSchema from './CompiledSchema';
import ChowError from '../error';
import { ChowOptions } from '..';

export default class CompiledParameterHeader {
  private compiledSchema: CompiledSchema;
  private headerSchema: SchemaObject = {
    type: 'object',
    properties: {},
    required: [],
    // All header properties should be a string, no?
    additionalProperties: { type: 'string' },
  };
  /**
   * If in is "header" and the name field is "Accept", "Content-Type" or "Authorization",
   * the parameter definition SHALL be ignored.
   */
  private ignoreHeaders = [
    'Accept',
    'Content-Type',
    'Authorization',
  ].map((header) => header.toLowerCase());

  constructor(parameters: ParameterObject[], options: Partial<ChowOptions>) {
    for (const parameter of parameters) {
      const headerName = parameter.name.toLowerCase();
      if (this.ignoreHeaders.includes(headerName)) {
        continue;
      }

      this.headerSchema.properties![headerName] = parameter.schema || {};

      if (parameter.required) {
        this.headerSchema.required!.push(headerName);
      }
    }

    this.compiledSchema = new CompiledSchema(
      this.headerSchema,
      options.headerAjvOptions
    );
  }

  /**
   * If there is no header passed in, we make it an empty object
   */
  public validate(value: any = {}) {
    try {
      this.compiledSchema.validate(value);
      return value;
    } catch (e) {
      throw new ChowError('Schema validation error', {
        in: 'header',
        rawErrors: e,
      });
    }
  }
}
