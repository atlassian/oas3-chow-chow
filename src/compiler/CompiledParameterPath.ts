import { ParameterObject, SchemaObject } from 'openapi3-ts';
import CompiledSchema from './CompiledSchema';
import ChowError from '../error';
import { ChowOptions } from '..';

export default class CompiledParameterPath {
  private compiledSchema: CompiledSchema;
  private pathSchema: SchemaObject = {
    type: 'object',
    properties: {},
    required: []
  };

  constructor(parameters: ParameterObject[], options: Partial<ChowOptions>) {
    for (const parameter of parameters) {
      this.pathSchema.properties![parameter.name] = parameter.schema || {};
      /**
       * All path parameters are required
       */
      this.pathSchema.required!.push(parameter.name);
    }

    /**
     * We want query to coerce to array if needed
     * For example:
     *   `/pets/123` will be valid against a schema with type=number even if `123` is string
     */
    this.compiledSchema = new CompiledSchema(this.pathSchema, { coerceTypes: true, ...(options.pathAjvOptions ? options.pathAjvOptions : {}) });
  }

  /**
   * If there is no path passed in, we make it an empty object
   */
  public validate(value: any) {
    try {
      const coercedValue = {...value};
      this.compiledSchema.validate(coercedValue);
      return coercedValue;
    } catch(e) {
      throw new ChowError('Schema validation error', { in: 'path', rawErrors: e });
    }
  }
}
