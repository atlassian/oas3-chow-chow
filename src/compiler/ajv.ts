import * as Ajv from 'ajv';

const options: Ajv.Options = {
  /**
   * Ignore following formats for now because they are not supported by AJV by default.
   * TODO: Add custom format supports for following formats.
   */
  unknownFormats: [ 'int32', 'int64', 'float', 'double', 'byte', 'binary', 'password' ],
  nullable: true,
  jsonPointers: true
}

export default function ajv(opts: Ajv.Options = {}) {
  return new Ajv({
    ...options,
    ...opts
  })
}
