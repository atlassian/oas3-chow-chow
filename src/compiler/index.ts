import {
  OpenAPIObject,
  PathItemObject,
} from "openapi3-ts";
import CompiledPath from "./CompiledPath";
import * as deref from "json-schema-deref-sync";
import { ChowOptions } from "..";

export interface RequestMeta {
  method: string;
  query?: any;
  header?: any;
  path?: any;
  cookie?: any;
  body?: any;
  operationId?: string;
}

export interface ResponseMeta {
  method: string;
  status: number;
  header?: any;
  body?: any;
}

export default function compile(oas: OpenAPIObject, options: Partial<ChowOptions>): CompiledPath[] {
  const document: OpenAPIObject = deref(oas, {failOnMissing: true});

  return Object.keys(document.paths).map((path: string) => {
    const pathItemObject: PathItemObject = document.paths[path];

    // TODO: support for base path
    return new CompiledPath(path, pathItemObject, options);
  });
}
