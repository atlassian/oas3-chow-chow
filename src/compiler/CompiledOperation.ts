import { OperationObject, ParameterObject, RequestBodyObject } from 'openapi3-ts';
import CompiledRequestBody from './CompiledRequestBody'
import CompiledResponse from './CompiledResponse';
import { RequestMeta, ResponseMeta } from '.';
import ChowError from '../error';
import CompiledParameterHeader from './CompiledParameterHeader';
import CompiledParameterQuery from './CompiledParameterQuery';
import CompiledParameterPath from './CompiledParameterPath';
import CompiledParameterCookie from './CompiledParameterCookie';
import CompiledMediaType from './CompiledMediaType';
import { ChowOptions } from '..';

export default class CompiledOperation {
  private header: ParameterObject[] = [];
  private compiledHeader: CompiledParameterHeader;
  private query: ParameterObject[] = [];
  private compiledQuery: CompiledParameterQuery;
  private path: ParameterObject[] = [];
  private compiledPath: CompiledParameterPath;
  private cookie: ParameterObject[] = [];
  private compiledCookie: CompiledParameterCookie;
  private body?: CompiledRequestBody;
  private operationId?: string;
  private response: {
    [key: string]: CompiledResponse;
  } = {};

  constructor(operation: OperationObject, options: Partial<ChowOptions>) {
    const parameters = !!operation.parameters ? [...operation.parameters] : []
    for (const parameter of parameters as ParameterObject[]) {
      switch(parameter.in) {
        case 'header':
          this.header.push(parameter);
          break;
        case 'query':
          this.query.push(parameter);
          break;
        case 'path':
          this.path.push(parameter);
          break;
        case 'cookie':
          this.cookie.push(parameter);
          break;
      }
    }
    this.compiledHeader = new CompiledParameterHeader(this.header, options);
    this.compiledQuery = new CompiledParameterQuery(this.query, options);
    this.compiledPath = new CompiledParameterPath(this.path, options);
    this.compiledCookie = new CompiledParameterCookie(this.cookie, options);

    if (operation.requestBody) {
      this.body = new CompiledRequestBody(operation.requestBody as RequestBodyObject);
    }

    this.operationId = operation.operationId;

    this.response = Object.keys(operation.responses).reduce((compiled: any, status: string) => {
      compiled[status] = new CompiledResponse(operation.responses[status]);
      return compiled;
    }, {});
  }

  public getDefinedRequestBodyContentType(): string[] {
    return this.body ? this.body.getDefinedContentTypes() : [];
  }

  public validateRequest(request: RequestMeta): RequestMeta {
    const header = this.compiledHeader.validate(request.header);
    const query = this.compiledQuery.validate(request.query);
    const path = this.compiledPath.validate(request.path);
    const cookie = this.compiledCookie.validate(request.cookie);

    let body;
    if (this.body) {
      const contentType = CompiledMediaType.extractMediaType(request.header && request.header['content-type']);
      body = this.body.validate(contentType, request.body);
    }

    return {
      method: request.method,
      operationId: request.operationId || this.operationId,
      header,
      query,
      path,
      cookie,
      body
    };
  }

  public validateResponse(response: ResponseMeta): ResponseMeta {
    const compiledResponse = this.response[response.status] || this.response['default'];
    if (compiledResponse) {
      return {...response, body: compiledResponse.validate(response)};
    } else {
      throw new ChowError('Unsupported Response Status Code', { in: 'response' });
    }
  }
}
