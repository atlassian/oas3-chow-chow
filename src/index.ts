import * as Ajv from 'ajv';
import { OpenAPIObject } from 'openapi3-ts';
import compile, { RequestMeta, ResponseMeta } from './compiler';
import CompiledPath from './compiler/CompiledPath';
import ChowError, {
  RequestValidationError,
  ResponseValidationError,
} from './error';
import CompiledOperation from './compiler/CompiledOperation';
import * as util from 'util';

/**
 * Export Errors so that consumers can use it to ditinguish different error type.
 */
export {
  default as ChowError,
  RequestValidationError,
  ResponseValidationError,
} from './error';

export interface ChowOptions {
  headerAjvOptions: Ajv.Options;
  cookieAjvOptions: Ajv.Options;
  pathAjvOptions: Ajv.Options;
  queryAjvOptions: Ajv.Options;
  requestBodyAjvOptions: Ajv.Options;
  responseBodyAjvOptions: Ajv.Options;
}

export default class ChowChow {
  private compiledPaths: CompiledPath[];
  private compiledOperationById: Map<string, CompiledOperation>;
  private deprecateValidateRequest = util.deprecate(
    this.validateRequestByPath.bind(this),
    'validateRequest() is now deprecated, please use validateRequestByPath or validateRequestByOperationId instead'
  );
  private deprecateValidateResponse = util.deprecate(
    this.validateResponseByPath.bind(this),
    'validateResponse() is now deprecated, please use validateResponseByPath or validateResponseByOperationId instead'
  );

  constructor(document: OpenAPIObject, options: Partial<ChowOptions> = {}) {
    const { compiledPaths, compiledOperationById } = compile(document, options);
    this.compiledPaths = compiledPaths;
    this.compiledOperationById = compiledOperationById;
  }

  validateRequest(path: string, request: RequestMeta & { method: string }) {
    return this.deprecateValidateRequest(path, request.method, request);
  }

  validateResponse(path: string, response: ResponseMeta & { method: string }) {
    return this.deprecateValidateResponse(path, response.method, response);
  }

  validateRequestByPath(path: string, method: string, request: RequestMeta) {
    try {
      const compiledPath = this.identifyCompiledPath(path);
      return compiledPath.validateRequest(path, method, request);
    } catch (err) {
      if (err instanceof ChowError) {
        throw new RequestValidationError(err.message, err.meta);
      } else {
        throw err;
      }
    }
  }

  validateResponseByPath(path: string, method: string, response: ResponseMeta) {
    try {
      const compiledPath = this.identifyCompiledPath(path);
      return compiledPath.validateResponse(method, response);
    } catch (err) {
      if (err instanceof ChowError) {
        throw new ResponseValidationError(err.message, err.meta);
      } else {
        throw err;
      }
    }
  }

  validateRequestByOperationId(operationId: string, request: RequestMeta) {
    const compiledOperation = this.compiledOperationById.get(operationId);

    if (!compiledOperation) {
      throw new ChowError(
        `No matches found for the given operationId - ${operationId}`,
        { in: 'request', code: 404 }
      );
    }

    try {
      return compiledOperation.validateRequest(request);
    } catch (err) {
      if (err instanceof ChowError) {
        throw new RequestValidationError(err.message, err.meta);
      } else {
        throw err;
      }
    }
  }

  validateResponseByOperationId(operationId: string, response: ResponseMeta) {
    const compiledOperation = this.compiledOperationById.get(operationId);

    if (!compiledOperation) {
      throw new ChowError(
        `No matches found for the given operationId - ${operationId}`,
        { in: 'response', code: 404 }
      );
    }

    try {
      return compiledOperation.validateResponse(response);
    } catch (err) {
      if (err instanceof ChowError) {
        throw new ResponseValidationError(err.message, err.meta);
      } else {
        throw err;
      }
    }
  }

  getDefinedRequestBodyContentType(path: string, method: string) {
    try {
      const compiledPath = this.identifyCompiledPath(path);
      return compiledPath.getDefinedRequestBodyContentType(method);
    } catch (err) {
      if (err instanceof ChowError) {
        return [];
      } else {
        throw err;
      }
    }
  }

  private identifyCompiledPath(path: string) {
    const compiledPath = this.compiledPaths.find((cp: CompiledPath) => {
      return cp.test(path);
    });

    if (!compiledPath) {
      throw new ChowError(`No matches found for the given path - ${path}`, {
        in: 'paths',
        code: 404,
      });
    }

    return compiledPath;
  }
}
