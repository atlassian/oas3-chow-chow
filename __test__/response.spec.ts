import ChowChow from '../src';
import ChowError, { ResponseValidationError } from '../src/error';
import {ResponseMeta} from '../src/compiler';

const fixture = require('./fixtures/response.json');

describe('Response', () => {
  let chowchow: ChowChow;

  beforeAll(() => {
    chowchow = new ChowChow(fixture);
  });

  it('should validate the response with status code', () => {
    const responseMeta: ResponseMeta = {
      method: 'get',
      status: 200,
      header: {
        'content-type': 'application/json'
      },
      body: [
        {
          id: 1,
          name: 'plum',
          readOnlyProp: '42',
          notWriteOnlyProp: '42'
        }
      ]
    };
    expect(chowchow.validateResponse('/pets/123', responseMeta)).toEqual(responseMeta);
  });

  it("should pass if a field that is nullable: true is null", () => {
      const responseMeta: ResponseMeta = {
          method: "get",
          status: 200,
          header: {
              "content-type": "application/json"
          },
          body: [
              {
                  id: 1,
                  name: "plum",
                  tag: null
              }
          ]
      };
      expect(chowchow.validateResponse("/pets/123", responseMeta)).toEqual(responseMeta);
  });

  it('should fail validation the response with writeOnly property', () => {
    const responseMeta: ResponseMeta = {
      method: 'get',
      status: 200,
      header: {
        'content-type': 'application/json'
      },
      body: [
        {
          id: 1,
          name: 'plum',
          writeOnlyProp: '42'
        }
      ]
    };
    expect(() => chowchow.validateResponse('/pets/123', responseMeta)).toThrow(ResponseValidationError);
  });

  it('should fall back to default if no status code is matched', () => {
    const responseMeta: ResponseMeta = {
      method: 'get',
      status: 500,
      header: {
        'content-type': 'application/json'
      },
      body: [
        {
          code: 500,
          message: 'something is wrong'
        }
      ]
    };
    expect(chowchow.validateResponse('/pets/123', responseMeta)).toEqual(responseMeta);
  });

  it('should throw error if path parameter fails schema check', () => {
    expect(() => {
      chowchow.validateRequest('/pets/abc', {
        method: 'get'
      });
    }).toThrowError(ChowError);
  });

  it('should fail on unsupported response media type', () => {
    expect(() => {
      chowchow.validateResponse('/pets/123', {
        method: 'get',
        status: 200,
        header: {
          'content-type': 'application/nonono'
        }
      });
    }).toThrow();
  })

  it('should ignore on empty response media type', () => {
    expect(() => {
      chowchow.validateResponse('/pets/123', {
        method: 'get',
        status: 204,
        header: {
          'content-type': ''
        }
      });
    }).not.toThrow();
  })

  it('should extract media type correctly in Content-Type header', () => {
    expect(() => {
      chowchow.validateResponse('/pets/123', {
        method: 'get',
        status: 200,
        header: {
          'content-type': 'application/json; charset=utf-8'
        }
      });
    }).toThrow();
  })

  it('should fail if response body is invalid', () => {
    expect(() => {
      chowchow.validateResponse('/pets/123', {
        method: 'get',
        status: 200,
        header: {
          'content-type': 'application/json'
        },
        body: [
          {
            id: 1
          }
        ]
      });
    }).toThrow();
  })

  it('should fail if response code does not match any', () => {
    expect(() => {
      chowchow.validateResponse('/no-default', {
        method: 'get',
        status: 400,
        header: {
          'content-type': 'application/json'
        }
      });
    }).toThrow();
  })

  it('should fail if response method is invalid', () => {
    expect(() => {
      chowchow.validateResponse('/no-default', {
        method: 'path',
        status: 400,
        header: {
          'content-type': 'application/json'
        }
      });
    }).toThrow();
  })

  it('should fail if header is invalid', () => {
    expect(() => {
      chowchow.validateResponse('/header', {
        method: 'get',
        status: 200,
        header: {
          'content-type': 'application/json',
          'version': [1, 2]
        }
      });
    }).toThrow();
  })

  it('should fail if required header is empty', () => {
    expect(() => {
      chowchow.validateResponse('/header', {
        method: 'get',
        status: 200
      });
    }).toThrow();
  })
});
