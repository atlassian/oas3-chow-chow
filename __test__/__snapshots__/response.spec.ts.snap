// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Response validateResponseByOperationId should fail if header value is invalid 1`] = `"ResponseValidationError: Schema validation error"`;

exports[`Response validateResponseByPath should fail if header value is invalid 1`] = `"ResponseValidationError: Schema validation error"`;
